docs/rfc/0000-json-schema-validation.md

```diff
@@ -0,0 +1,272 @@
+  return { path, validation };
+}
+
+const results = Promise.all(yamls.map(f => validateFile(f.path)));
+results.then(function(rs) {
+  for (const { path, validation } of rs) {
+    if (validation == 'ok') {
+      log(`${path}: ok`);
+      return;
+    }
+    // otherwise, it's a list of errors
+    for (err of validation) {
+      log(`${path}: error: ${err}`);
+    }
+  }
+});
+```
+
+## Motivation
+
+`jk` and its libraries provide various means for generating
+configuration, but there is in general no guarantee that the output is
+usable. For that reason, to become a more rounded tool, validation
+should be considered a core capability for `jk`.
+
+One very general-purpose means of validation is by checking against a
+data model or schema. In the case of JSON-serialisable JavaScript
+objects -- which is effectively the data model of `jk` -- a reasonable
+choice of schema language is JSON Schema. Schemas using JSON Schema
+(and the closely related OpenAPI data model, more on which see below)
+are available for lots of APIs, and it has support in tooling e.g.,
+[VSCode](https://code.visualstudio.com/docs/languages/json#_json-schemas-and-settings).
+
+Building in schema validation means libraries are free to rely on it,
+encouraging more widespread validation.
+
+## Design
+
+### Discussion
+
+**Synchronisation**
+
+The expectation might be that validation is synchronous, since it is
+used mostwhere as a predicate. There's good reasons why it should be
+asynchronous, though:
+
+ - validating with a file means a file read, and those are
+   asynchronous, so the precedent exists;
+ - resolving references may involve network requests, and those are
+   naturally represented as asynchronous.
+
+**Resolving schemas**
+
+ - `$schema` values will likely refer to public URIs (but this RFC
+   says nothing about respecting `$schema`; that could follow in
+   another RFC, perhaps).
+ - schemas will often refer to other schemas with `$ref: <json
+   pointer>`, which can come from the same file, or from a file
+   relative to the current file, or a URI (though it's not required to
+   be a network locator, i.e., it's up to the library how to process
+   it)
+ - libraries will want to include their schemata in the distributed
+   package, so will want to be able to load them as a resource.
+
+**Caching**
+
+Usually, caching is strictly an optimisation, and as such it is not
+necessary to discuss it by way of specification. But it's worth
+mentioning that Go libraries for schema checking tend to come with a
+cache for external schemas; and some schemas will be used extensively
+during a run (e.g., the schema for a Kubernetes Deployment). It would
+be good to work in sympathy with any caching in the library, where
+possible.
+
+### API in JavaScript
+
+The result of a schema check is either that everything was OK, or that
+there were some specific problems.
+
+```typescript
+type ValidationResult = 'ok' | string[];
```

Comment:
I think we should allow for more context in the error:
- start and end (line, column) tuples of the error
- jsonpointer of the error
- error string

Of course it will depend on the underlying library if we can provide those easily but worth the try of having more structured data (or at least the potential for more structured data in the api.

So, instead of a `[]string` I’d put an array of structs, maybe marking fields other than the error message optional.

Reply:
Good call; here and everywhere I could make the validation return value

```
interface ValidationError {
  message: string;
  pointer?: string;
  location: [number, number];
}
```

Reply:
`[{line, col}, {line, col}]` has the benefit to be able to express multi-line error contexts as well as pin-pointing a word on a line. It's also compatible with the github check API when you are reporting errors on files that are in a git repo so github can display a nice highlight on the problematic part.

Would go for `[{line, col}, {line, col}]`. Maybe make the second part optional if the first one is provided, underlying libraries may not provide that information? Could be:

```ts
interface Location {
  line: number;
  col: number;
}

interface ValidationError {
  message: string;
  pointer?: string;
  start?: Location;
  end?: Location;
}
```

Reply:
Adopted that last one